crates/precompile/src/crypto_provider/mod.rs

```diff
@@ -0,0 +1,618 @@
+    /// * `b` - Second G2 point as a tuple (x0, x1, y0, y1) with each component being bls12_381::FP_LENGTH bytes
+    ///
+    /// # Returns
+    /// The sum as bls12_381::G2_LENGTH bytes (unpadded), or an error if points are invalid.
+    fn bls12_381_g2_add(
+        &self,
+        a: bls12_381::G2Point,
+        b: bls12_381::G2Point,
+    ) -> Result<[u8; bls12_381::G2_LENGTH], PrecompileError>;
+
+    /// BLS12-381 G1 multi-scalar multiplication.
+    ///
+    /// # Arguments
+    /// * `points_scalars` - Iterator of (point, scalar) pairs where:
+    ///   - Points are G1Point tuples: (x, y) with each component being bls12_381::FP_LENGTH bytes
+    ///   - Scalars are bls12_381::SCALAR_LENGTH bytes each
+    ///
+    /// # Returns
+    /// The result as bls12_381::G1_LENGTH bytes (unpadded), or an error if inputs are invalid.
+    fn bls12_381_g1_msm(
+        &self,
+        points_scalars: Box<
+            dyn Iterator<Item = Result<bls12_381::G1PointScalarPair, PrecompileError>> + '_,
+        >,
+    ) -> Result<[u8; bls12_381::G1_LENGTH], PrecompileError>;
+
+    /// BLS12-381 G2 multi-scalar multiplication.
+    ///
+    /// # Arguments
+    /// * `points_scalars` - Iterator of (point, scalar) pairs where:
+    ///   - Points are G2Point tuples: (x0, x1, y0, y1) with each component being bls12_381::FP_LENGTH bytes
+    ///   - Scalars are bls12_381::SCALAR_LENGTH bytes each
+    ///
+    /// # Returns
+    /// The result as bls12_381::G2_LENGTH bytes (unpadded), or an error if inputs are invalid.
+    fn bls12_381_g2_msm(
+        &self,
+        points_scalars: Box<
+            dyn Iterator<Item = Result<bls12_381::G2PointScalarPair, PrecompileError>> + '_,
+        >,
+    ) -> Result<[u8; bls12_381::G2_LENGTH], PrecompileError>;
+
+    /// BLS12-381 pairing check.
+    ///
+    /// # Arguments
+    /// * `pairs` - Vector of PairingPair (G1, G2) point pairs for pairing check
+    ///
+    /// # Returns
+    /// `true` if the pairing check passes, `false` otherwise.
+    fn bls12_381_pairing(&self, pairs: &[bls12_381::PairingPair]) -> Result<bool, PrecompileError>;
+
+    /// BLS12-381 map field element to G1.
+    ///
+    /// # Arguments
+    /// * `fp` - Field element as bls12_381::FP_LENGTH bytes
+    ///
+    /// # Returns
+    /// The mapped G1 point as bls12_381::G1_LENGTH bytes (unpadded).
+    fn bls12_381_map_fp_to_g1(
+        &self,
+        fp: &[u8; bls12_381::FP_LENGTH],
+    ) -> Result<[u8; bls12_381::G1_LENGTH], PrecompileError>;
+
+    /// BLS12-381 map field element to G2.
+    ///
+    /// # Arguments
+    /// * `fp2` - Fp2 element as two bls12_381::FP_LENGTH-byte field elements
+    ///
+    /// # Returns
+    /// The mapped G2 point as bls12_381::G2_LENGTH bytes (unpadded).
+    fn bls12_381_map_fp2_to_g2(
+        &self,
+        fp2_x: &[u8; bls12_381::FP_LENGTH],
+        fp2_y: &[u8; bls12_381::FP_LENGTH],
+    ) -> Result<[u8; bls12_381::G2_LENGTH], PrecompileError>;
```

Comment:
Technically speaking, putting PrecompileError here and in the APIs for the cryptography code is leaking an abstraction, but I think it can be cleaned up after